### PR TITLE
[022] add internships relations

### PR DESCRIPTION
**Pull request by Maksim-maevsky**
- added enums for format and skill 
- added tables user-skill and internship skill
- added initialization scripts for  user-skill and internship skill

### CODE REVIEW

**Review by ShirshovAlex — DISMISSED** [review]

**Review by velichko1882 — DISMISSED** [review]

File: src/main/java/by/exadel/internship/entity/Internship.java [hunk 1]
```diff
@@ -55,11 +53,17 @@
...
 
-//    private List<FormFullDTO> formList;
-//    private List<UserDTO> techList;
-//    private List<UserDTO> adminList;
-//    private List<LocationDTO> countryList;
-//    private InternshipFormat InternshipFormat;
-//    private List<Technology> technologyList;
+    @Enumerated(EnumType.STRING)
+    @Column(name = "inship_format_name", nullable = false)
+    private InternshipFormat InternshipFormat;
```
Comment (mortal-gpt) [inline src/main/java/by/exadel/internship/entity/Internship.java]:
```suggestion
    private InternshipFormat internshipFormat;
```

File: src/main/java/by/exadel/internship/entity/Internship.java [hunk 2]
```diff
@@ -55,11 +53,17 @@
...
+
+    @ManyToOne
+    @JoinColumn(name="inship_technology_name", nullable=false)
+    private Technology technology;
```
Comment (mortal-gpt) [inline src/main/java/by/exadel/internship/entity/Internship.java]: Just note: Did you investigate posibility to use  @ElementCollection?

File: src/main/resources/application.yaml [hunk 3]
```diff
@@ -2,20 +2,23 @@ spring:
   jpa:
     show-sql: true
     hibernate:
-      ddl-auto: validate
+      ddl-auto: update
```
Comment (mortal-gpt) [inline src/main/resources/application.yaml]: revert configuration changes please

File: src/main/java/by/exadel/internship/entity/Internship.java [hunk 4]
```diff
@@ -55,11 +53,17 @@
...
     private LocalDate registrationEndDate;
 
-//    private List<FormFullDTO> formList;
-//    private List<UserDTO> techList;
-//    private List<UserDTO> adminList;
-//    private List<LocationDTO> countryList;
-//    private InternshipFormat InternshipFormat;
-//    private List<Technology> technologyList;
+    @Enumerated(EnumType.STRING)
+    @Column(name = "inship_format_name", nullable = false)
```
Comment (kzaretskaya) [inline src/main/java/by/exadel/internship/entity/Internship.java]: Why `nullable = false`? I think when admin create internship it's possible that format can not be approved yet.

File: src/main/resources/application.yaml [hunk 5]
```diff
@@ -2,16 +2,17 @@ spring:
...
   datasource:
     driver-class-name: org.postgresql.Driver
-    url: ${SPRING_DATASOURCE_URL}
-    password: ${SPRING_DATASOURSE_PASSWORD}
-    username: ${SPRING_DATASOURSE_USERNAME}
+    url: jdbc:postgresql://localhost:5432/postgres
```
Comment (kzaretskaya) [inline src/main/resources/application.yaml]: revert to previous state

File: src/main/java/by/exadel/internship/entity/user/User.java [hunk 6]
```diff
@@ -1,20 +1,20 @@
...
 import lombok.Data;
 import lombok.NoArgsConstructor;
+
 import javax.persistence.*;
-import java.util.List;
 import java.util.UUID;
 
 @Entity
-@Table(name = "user")
+@Table(name = "user", schema = "public")
```
Comment (kzaretskaya) [inline src/main/java/by/exadel/internship/entity/user/User.java]: what purposes do you specify schema for?

File: src/main/java/by/exadel/internship/entity/Internship.java (hunk 1 above)
Comment (Maksim-maevsky) [reply src/main/java/by/exadel/internship/entity/Internship.java]: done

File: src/main/resources/application.yaml (hunk 3 above)
Comment (Maksim-maevsky) [reply src/main/resources/application.yaml]: done

File: src/main/java/by/exadel/internship/entity/user/User.java (hunk 6 above)
Comment (Maksim-maevsky) [reply src/main/java/by/exadel/internship/entity/user/User.java]: IT WAS Grammar exception from hibrnate, and it was pretty strange, because everything was right, and schema was decision from net about this problem.

Comment (Fenydude) [reply src/main/java/by/exadel/internship/entity/user/User.java]: I think that because postgres has another user table and schema = "public" resolves that conflict

File: src/main/resources/application.yaml (hunk 5 above)
Comment (Maksim-maevsky) [reply src/main/resources/application.yaml]: done

File: src/main/java/by/exadel/internship/entity/user/User.java (hunk 6 above)
Comment (Maksim-maevsky) [reply src/main/java/by/exadel/internship/entity/user/User.java]: yes, but we took desicion, how i remember, about singular naming. @kzaretskaya  maybe we should change this?

File: src/main/java/by/exadel/internship/entity/Internship.java (hunk 2 above)
Comment (Maksim-maevsky) [reply src/main/java/by/exadel/internship/entity/Internship.java]: done 
image

File: src/main/java/by/exadel/internship/entity/Internship.java (hunk 4 above)
Comment (Maksim-maevsky) [reply src/main/java/by/exadel/internship/entity/Internship.java]: I am going to text  Zanna today about it with other questions.

File: src/main/java/by/exadel/internship/entity/Internship.java (hunk 2 above)
Comment (Maksim-maevsky) [reply src/main/java/by/exadel/internship/entity/Internship.java]: image

**Review by Fenydude — DISMISSED** [review]

File: src/main/java/by/exadel/internship/entity/User.java [hunk 7]
```diff
@@ -38,10 +40,13 @@
     @Enumerated(EnumType.STRING)
     private UserRole userRole;
 
-    //    @Column(name = "technology")
-//    private List<Technology> techTechnology;
+    @Column(name="skill_name", nullable=false)
+    @Enumerated(EnumType.STRING)
+    @ElementCollection(fetch = FetchType.EAGER)
+    @CollectionTable(name = "user_skill", joinColumns = @JoinColumn(name = "u_s_u_id"))
```
Comment (mortal-gpt) [inline src/main/java/by/exadel/internship/entity/User.java]:
```suggestion
    @CollectionTable(name = "user_skill", joinColumns = @JoinColumn(name = "us_u_id"))
```

File: src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+      author: maksim_maevsky
+      changes:
+        - createTable:
+            tableName: internship_skill
+            columns:
+              - column:
+                  name: skill_name
+                  type: SKILL
+                  constraints:
+                    nullable: false
```
Comment (mortal-gpt) [inline src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml:19]:
```suggestion
                    nullable: true
```

skill_name can't exist without is_inship_id

File: src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+              - column:
+                  name: i_s_inship_id
```
Comment (mortal-gpt) [inline src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml]: name: is_inship_id

File: src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+  - changeSet:
+      id: 12042021-add-type-user-skill
+      author: maksim_maevsky
+      changes:
+        - createTable:
+            tableName: user_skill
+            columns:
+              - column:
+                  name: u_s_u_id
```
Comment (mortal-gpt) [inline src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml]:
```suggestion
                  name: us_u_id
```

File: src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+                  type: UUID
+                  constraints:
+                    nullable: true
+              - column:
+                  name: skill_name
```
Comment (mortal-gpt) [inline src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml]:
```suggestion
                  name: us_name
```

File: src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+      author: maksim_maevsky
+      changes:
+        - createTable:
+            tableName: user_skill
+            columns:
+              - column:
+                  name: u_s_u_id
+                  type: UUID
+                  constraints:
+                    nullable: true
```
Comment (mortal-gpt) [inline src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml]:
```suggestion
                    nullable: false
```

Comment (mortal-gpt) [reply src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml]: us_u_id doesn't exist without skill_name

File: src/main/java/by/exadel/internship/entity/User.java (hunk 7 above)
Comment (Maksim-maevsky) [reply src/main/java/by/exadel/internship/entity/User.java]: done

File: src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml (hunk 8 above)
Comment (Maksim-maevsky) [reply src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml:19]: done

File: src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml (hunk 9 above)
Comment (Maksim-maevsky) [reply src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml]: done

File: src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml (hunk 10 above)
Comment (Maksim-maevsky) [reply src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml]: done

File: src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml (hunk 11 above)
Comment (Maksim-maevsky) [reply src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml]: done

File: src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml (hunk 12 above)
Comment (Maksim-maevsky) [reply src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml]: done

File: src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml (hunk 8 above)
Comment (Maksim-maevsky) [reply src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml:19]: maybe false is right?

File: src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+            constraintName: fk_internship_skill
+            referencedColumnNames: inship_id
+            referencedTableName: internship
+
+  - changeSet:
+      id: 12042021-add-type-user-skill
+      author: maksim_maevsky
+      changes:
+        - createTable:
+            tableName: user_skill
```
Comment (kzaretskaya) [inline src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml:51]: Please, add to both table uniqie constraint to avoid set the same skill to one user/internship.

File: src/main/java/by/exadel/internship/entity/Internship.java [hunk 14]
```diff
@@ -55,11 +55,20 @@
...
 
-//    private List<FormFullDTO> formList;
-//    private List<UserDTO> techList;
-//    private List<UserDTO> adminList;
-//    private List<LocationDTO> countryList;
-//    private InternshipFormat InternshipFormat;
-//    private List<Technology> technologyList;
+    @Column(name = "inship_format_name", nullable = false)
+    @Enumerated(EnumType.STRING)
+    @Type(type = "by.exadel.internship.mapper.enum_mapper.EnumTypePostgreSQL")
```
Comment (kzaretskaya) [inline src/main/java/by/exadel/internship/entity/Internship.java:59]: why do you specify type here?

File: src/main/java/by/exadel/internship/entity/Location.java [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import java.util.UUID;
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class Location {
```
Comment (kzaretskaya) [inline src/main/java/by/exadel/internship/entity/Location.java]: Remove please the class from this PR. It'll be added in separate PR and with all appropriate annotation.

File: src/main/java/by/exadel/internship/entity/Internship.java (hunk 14 above)
Comment (Maksim-maevsky) [reply src/main/java/by/exadel/internship/entity/Internship.java:59]: it is for the future , when we will add some  Inrernship.

File: src/main/java/by/exadel/internship/entity/Location.java (hunk 15 above)
Comment (Maksim-maevsky) [reply src/main/java/by/exadel/internship/entity/Location.java]: done

File: src/main/java/by/exadel/internship/entity/user/User.java (hunk 6 above)
Comment (kzaretskaya) [reply src/main/java/by/exadel/internship/entity/user/User.java]: What will be changed? to plural naming? no. This entity is already in branch development, how does it work there? If problem appeared on your branch it need to figure out cause.

File: src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml (hunk 13 above)
Comment (Maksim-maevsky) [reply src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml:51]: done
image

File: src/main/java/by/exadel/internship/entity/user/User.java (hunk 6 above)
Comment (Maksim-maevsky) [reply src/main/java/by/exadel/internship/entity/user/User.java]: image
this is a working option also

Comment (Maksim-maevsky) [reply src/main/java/by/exadel/internship/entity/user/User.java]: https://stackoverflow.com/questions/22256124/cannot-create-a-database-table-named-user-in-postgresql

user is a reserved word and it's usually not a good idea use reserved words for identifiers (tables, columns).

**Review by Fenydude — DISMISSED** [review]

File: src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml (hunk 8 above)
Comment (SergeyKoval) [reply src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml:19]: Looks like false is correct

File: src/main/java/by/exadel/internship/entity/User.java [hunk 16]
```diff
@@ -38,10 +40,13 @@
     @Enumerated(EnumType.STRING)
     private UserRole userRole;
 
-    //    @Column(name = "technology")
-//    private List<Technology> techTechnology;
+    @Column(name="us_name", nullable=false)
+    @Enumerated(EnumType.STRING)
+    @ElementCollection(fetch = FetchType.EAGER)
```
Comment (SergeyKoval) [inline src/main/java/by/exadel/internship/entity/User.java]: Do you really need EAGER here? 
I suggest use LAZY by default as much as possible. 

Here I expect a lot of scenarios where this list of skills will be redundant. For example to assign user for the interview you'll need to get all users with required skill. Getting all skills for each user will be redundant.

**Review by SergeyKoval — CHANGES_REQUESTED** [review]

File: src/main/java/by/exadel/internship/entity/User.java [hunk 17]
```diff
@@ -1,20 +1,22 @@
...
 import lombok.NoArgsConstructor;
+
 import javax.persistence.*;
-import java.util.List;
+import java.util.Set;
 import java.util.UUID;
 
 @Entity
-@Table(name = "user")
+@Table(name="\"user\"")
```
Comment (kzaretskaya) [inline src/main/java/by/exadel/internship/entity/User.java:14]: I think it's not a good idea to wrap with quotes. This problem had to be appeared when the entity was created (probably it's not tested).
My suggestion to rename table to `user_detail`.
@viktarkolbik @SergeyKoval what do you think?

File: src/main/java/by/exadel/internship/entity/user/User.java (hunk 6 above)
Comment (kzaretskaya) [reply src/main/java/by/exadel/internship/entity/user/User.java]: I think it's not a good idea to wrap with quotes. This problem had to be appeared when the entity was created (probably it's not tested).
My suggestion to rename table to user_detail.
@viktarkolbik @SergeyKoval what do you think?

Comment (SergeyKoval) [reply src/main/java/by/exadel/internship/entity/user/User.java]: Agree with @kzaretskaya

File: src/main/java/by/exadel/internship/entity/Internship.java (hunk 4 above)
Comment (Maksim-maevsky) [reply src/main/java/by/exadel/internship/entity/Internship.java]: done like nullable to Internship and  db

File: src/main/java/by/exadel/internship/entity/user/User.java (hunk 6 above)
Comment (Maksim-maevsky) [reply src/main/java/by/exadel/internship/entity/user/User.java]: done 
image

File: src/main/java/by/exadel/internship/entity/User.java (hunk 16 above)
Comment (Maksim-maevsky) [reply src/main/java/by/exadel/internship/entity/User.java]: i got it. Done.

File: src/main/java/by/exadel/internship/entity/Internship.java (hunk 14 above)
Comment (Maksim-maevsky) [reply src/main/java/by/exadel/internship/entity/Internship.java:59]: we will think about how we could use it from one place, thanks

**Review by kzaretskaya — APPROVED** [review]

**Review by OlgaEvdokimova — APPROVED** [review]

**Review by SergeyKoval — APPROVED** [review]
Don't forget to reduce number of sql queriws
